### ISSUE TITLE
Typescript shorthand style function type-checking

### ISSUE THREAD

**Issue by rnarcos** [styled-components/styled-components#3711]
## Environment
styled-components: 5.2.1
typescript: 4.5.4
react (and react-dom): 16.13.1

## Reproduction
https://codesandbox.io/s/styled-components-typescript-forked-czie1n?file=/src/App.tsx

## Steps to reproduce
1. Create a style function that returns a `css`, which extends a predefined interface;
```typescript
interface StyleCustomization {
  marginTop: string;
  marginBottom: string;
}

interface WithStyleCustomization {
  styleCustomization: StyleCustomization;
}

function customizationStyle<Props extends WithStyleCustomization>(
  props: Props
) {
  const { styleCustomization } = props;
  const { marginTop, marginBottom } = styleCustomization;

  return css`
    margin-top: ${marginTop};
    margin-bottom: ${marginBottom};
  `;
}
```
2. Use style function with shorthand syntax (`${customizationStyle};`, instead of `${(props) => customizationStyle(props)};`

## Expected Behavior
Typescript should warn for mis-use of `customizationStyle` function, on non-shorthand and shorthand syntaxes.

## Actual Behavior
Typescript only warn for mis-use of `customizationStyle` function on non-shorthand syntax.
<img width="842" alt="image" src="https://user-images.githubusercontent.com/35005829/160664728-0973cc9c-2bca-4230-bf35-19209ea2b07a.png">

Also, when using shorthand syntax, I notice that if I try to use a non-shorthand syntax, I receive the intellisense for the appropriate type definition that I was expecting.
<img width="590" alt="image" src="https://user-images.githubusercontent.com/35005829/160664963-f0b122e1-bc60-48a3-b463-32a9e1e28a8a.png">

I'm not entirely sure if this is something related to typescript itself, or styled-components. Please advise if that is the case.

[]s,

**Comment by quantizor** [comment]
Closing — this is a usage question rather than a bug report. For help, please use GitHub Discussions or Stack Overflow.
